### PR TITLE
Python camerad for webcam

### PR DESCRIPTION
**Pull request by bongbui321**
For [#24590](https://github.com/commaai/openpilot/issues/24590)

### CODE REVIEW

File: tools/webcam/camerad.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import threading
+import os
+
+from cereal.visionipc import VisionIpcServer, VisionStreamType
+from cereal import messaging
+
+from openpilot.tools.webcam.camera import Camera
+
+YUV_BUFFER_COUNT = 20 # same as c++ of camerad
+CAMERA_ROAD_ID, CAMERA_DRIVER_ID, CAMERA_WIDE_ID = 0, 1, 2
```
Comment (adeebshihadeh) [inline tools/webcam/camerad.py:11]: let's make these environment variables so they can be changed

File: tools/webcam/camerad.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  def cam_send_yuv_driver(self, yuv):
+    self._send_yuv(yuv, self.frame_wide_id, 'driverCameraState', VisionStreamType.VISION_STREAM_DRIVER)
+    self.frame_driver_id += 1
+
+  def cam_send_yuv_wide_road(self, yuv):
+    self._send_yuv(yuv, self.frame_wide_id, 'wideRoadCameraState', VisionStreamType.VISION_STREAM_WIDE_ROAD)
+    self.frame_wide_id += 1
+
+  def run(self):
+    self.t_cam_road = threading.Thread(target=Camerad.daemon_alive, args=(self.cam_road, self.cam_send_yuv_road))
```
Comment (adeebshihadeh) [inline tools/webcam/camerad.py:66]: parameterize all the camera stuff such that all the lines aren't being copy/pasted each time for road/driver/wide

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]
Can you also update the webcam README?

As verification, can you post a segment with this running? Just run camerad, loggerd, and encoderd and upload the rlog and .hevc files.

**Comment by bongbui321** [comment]
I don't have a comma device so I don't think I can upload to comma connect? I have ran and logged a minute of driving using a Simulated car similar to that in Simulation and use a pre-recorded driving video for camerad. Does that work? I can send you the .hevc and rlog files if thats okay because I can't upload straight to github since it doesn't support hevc and log files

File: tools/webcam/camerad.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      cam_thread.start()
+      self.camera_threads.append(cam_thread)
+
+  def join_camera_threads(self):
+    for thread in self.camera_threads:
+      thread.join()
+
+  def run(self):
+    self.start_camera_threads()
+    self.join_camera_threads()
```
Comment (adeebshihadeh) [inline tools/webcam/camerad.py:82]: why are these split into different functions? just put them in here directly

File: tools/webcam/camerad.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    function()
+    rk.keep_time()
+
+class Camerad:
+  def __init__(self):
+    self.pm = messaging.PubMaster(['roadCameraState', 'driverCameraState', 'wideRoadCameraState'])
+    self.vipc_server = VisionIpcServer("camerad")
+
+    self.dual_camera = bool(int(os.getenv("DUAL","0")))
+    self.cameras, self.camera_threads = [], [] # ORDER: road_cam, driver_cam, wide_cam
```
Comment (adeebshihadeh) [inline tools/webcam/camerad.py:29]: why does the order matter? it shouldn't

File: tools/webcam/camerad.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        if not self.dual_camera:
+          break
+        cam = Camera(cam_type, stream_type, int(os.getenv("CAMERA_WIDE_ID", "2")))
+      assert cam.cap.isOpened(), f"Can't find {cam_type}"
+      self.cameras.append(cam)
+      self.vipc_server.create_buffers(stream_type, YUV_BUFFER_COUNT, False, cam.W, cam.H)
+
+    self.vipc_server.start_listener()
+
+  #from sim
```
Comment (adeebshihadeh) [inline tools/webcam/camerad.py:46]: unnecessary comment. this is a simple function

File: tools/webcam/camerad.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    msg = {
+      "frameId": frame_id,
+      "transform": [1.0, 0.0, 0.0,
+                    0.0, 1.0, 0.0,
+                    0.0, 0.0, 1.0]
+    }
+    setattr(dat, pub_type, msg)
+    self.pm.send(pub_type, dat)
+
+  def daemon_alive(self, cam):
```
Comment (adeebshihadeh) [inline tools/webcam/camerad.py:60]: give this a better name like "camera_thread" or "camera_runner"

File: tools/webcam/start_camerad.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+export BLOCK="${BLOCK},camerad" #block camerad in terminal that calls manager.py
+export DUAL="0"
+export USE_WEBCAM="1" #export this in shell that calls manager.py
+export YUV_BUFFER_COUNT="20"
```
Comment (adeebshihadeh) [inline tools/webcam/start_camerad.sh:6]: no need to parameterize this. 20 should always be fine

File: tools/webcam/start_camerad.sh [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+#Change camera index according to your setting
```
Comment (adeebshihadeh) [inline tools/webcam/start_camerad.sh:8]: leave one space between the # and comment

File: tools/webcam/camerad.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+from openpilot.tools.webcam.camera import Camera
+from openpilot.common.realtime import Ratekeeper
+
+
+YUV_BUFFER_COUNT = int(os.getenv("YUV_BUFFER_COUNT", "20"))
+CAMERA_TYPE_STATE = {"roadCameraState":VisionStreamType.VISION_STREAM_ROAD,
+                    "driverCameraState":VisionStreamType.VISION_STREAM_DRIVER,
+                    "wideRoadCameraState":VisionStreamType.VISION_STREAM_WIDE_ROAD}
+
+def rk_loop(function, hz, exit_event: threading.Event):
```
Comment (adeebshihadeh) [inline tools/webcam/camerad.py:17]: this is unused

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]
Correct, you can't upload to connect, but that's not necessary. And no, a "simulated" camera won't work. We need a log generated from exactly this branch, not a modified one.

Just startup camerad, loggerd, and encoderd, then upload the rlog and hevc files somewhere like Google Drive.

File: tools/webcam/camerad.py (hunk 4 above)
Comment (bongbui321) [reply tools/webcam/camerad.py:29]: It doesn't it was for me in the beginning. I will delete it

**Comment by bongbui321** [comment]
[Link to hevc and rlog](https://drive.google.com/file/d/1XdKdNgkJJaFkGxO3B6VLsXVmC7VRCqlj/view?usp=drive_link)

In this log, I just change the camera_id from integer to a video file path in my machine, everything else is the same.

**Comment by sivaariram** [comment]
Are there any version available to use openpilot with webcam and PC ?

**Comment by MikeBusuttil** [comment]
> Are there any version available to use openpilot with webcam and PC ?

will find out and let you know @sivaariram

**Comment by MikeBusuttil** [comment]
Yo @bongbui321 FYI [this line in the README](https://github.com/commaai/openpilot/blob/57fc4f79d16c132d526a66d603fd9100804569d1/tools/webcam/README.md?plain=1#L27) references the old `camera_webcam.cc` that no longer exists.  I'll make an issue for it

**Pull request merged**